src/main/java/com/salesforce/functions/jvm/sdk/data/bulk/DataTableBuilder.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2022, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+package com.salesforce.functions.jvm.sdk.data.bulk;
+
+import com.salesforce.functions.jvm.sdk.data.Record;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+
+/** A builder for {@link DataTable} objects. */
+public interface DataTableBuilder {
+  /**
+   * Adds a new row, using the given map as the source for the field values. Values at keys which
+   * are not a field in the data table will be ignored. Missing values in the given map will result
+   * in null values being written for the respective field.
+   *
+   * @param rowData The data for the row.
+   * @return This {@link DataTableBuilder} instance to allow method chaining.
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  DataTableBuilder addRow(Map<String, String> rowData);
+
+  /**
+   * Adds a new row, using the given values as the source for the field values. Values are mapped by
+   * position, so if the first field name is "zebra", the value for this field has to come first in
+   * the argument list.
+   *
+   * @param rowData The data for the row.
+   * @return This {@link DataTableBuilder} instance to allow method chaining.
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  DataTableBuilder addRow(String... rowData);
+
+  /**
+   * Adds a new row, using the given {@link Record} as the source for the field values. Fields in
+   * the record which are not a field in the data table will be ignored. Missing fields in the given
+   * record will result in null values being written for the respective field in the data table.
+   *
+   * @param record The record to source the data for the row from.
+   * @return This {@link DataTableBuilder} instance to allow method chaining.
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  DataTableBuilder addRow(Record record);
```

Comment:
wait, wouldn't this go against your separation of concerns?  if the data table builder is aware of `Record` and how to extract data from such an object for ingestion, shouldn't it also stand to reason that it knows how to construct tables that are valid for ingestion (i.e.; automatically split for the user)?  

i would argue that this is not, in fact, a primitive data table builder but an ingest-specific data table builder that produces `DataTable` instances that are always valid for ingestion.